adrs/0541531435/README.md

```diff
@@ -0,0 +1,118 @@
+`tsconfig.json` files and many other files and directories.
+
+Subjectively speaking, they add noise and hard to navigate or concentrate when
+put all your files at the root of the project.
+
+Furthermore,
+
+* Where do I suppose to put my test files? What are the differences between X and
+Y directories?
+
+* How to structure the project when I have multiple testing frameworks, or
+multiple documentation tools?
+
+This ADR is an attend to follow the steps of other ecosystems like Elixir,
+Ruby or Rust came to a consensus about this topic.
+
+### About compilations
+
+We should tell what requires compilation in the project, which is extremely
+useful in library projects since you should be aware of such detail to consume
+the project. Or in cases that you would like to run the project in production.
+
+Having a directory that reflects such information is crucial to mitigate mistakes
+in application deployments, or misuse of libraries.
+
+### About test frameworks
+
+Sometimes you may have multiple test frameworks for your project, most likely
+for `end-to-end`, `integration`, or `unit` tests, but not limited to those
+concepts. Therefore, people have tried to come up with `./tests/integration`,
+`./tests/unit`, `./tests/end-to-end`.
+
+Although that may be a clean way to structure your tests, the problems arrive
+when you try to articulate the differences between those categories without
+getting into an in-depth conversation about X vs. Y and end up with unclear
+boundaries, relying too much on the programmer's experiences to articulate the
+most optimal definition or decision.
+
+Additionally, if you try to migrate your project to another testing tool that
+falls into one of those possible categories, you will mingle tests that are
+run differently depending on the underline tool, and that may be really
+important in some cases.
+
+### About documentation
+
+Documentation tools are amazingly diverse in the ecosystem; such tools require
+different configurations and code structure, therefore mingle tools may cause
+incompatible situations, or confusing documentation source code.
+
+Also, it is unclear today what should I do if I have multiple documentation
+websites (internal or public as an example), and how to manage them.
+
+## Resolution
+
+* You **must** follow the following directory structures:
+
+  ```text
+  .
+  ├── package.json
+  ├── bin
+  ├── dist
+  ├── docs
+  │   ├── storybook
+  │   └── website
+  │   └── website-next
+  │   └── [another website or tool]
+  ├── examples
+  ├── lib
+  ├── src
+  └── tests
+      ├── cypress
+      └── jest
+      └── puppeteer
+      └── [another testing tool]
+  ```
+
+* In the `./tests` directory, you **must** use the tool's name to reflect the
+  tool used to run the tests.
+* You **must** place `storybook` documentation under `./docs`.
+* In the `./docs` directory, you **must** separate documentation based on the
+  underline tool or context (for example, **internal** or **public**).
```

Comment:
This is not an ADR I would adopt on my projects for these highlighted bullet points.

I've worked on projects that follow this ADR and projects that put test files & story files next to the file they're testing & story booking.  I strongly prefer the latter.  

When these files are next to each other, it is significantly easier to navigate and see which files have spec coverage and stories.  Each directory becomes its own "workspace."  This is more in line with the development flow; we build, test, and document in tandem.

Reply:
Although I agree with you in terms of "it is easier to navigate"  the added complexity, and noise in production source code it is not worth it.

Worth pointing out one common mistake that people make is to create a one-to-one relationship for testing files, storybook, and so on, I would subjectively say that most of the time you are just adding entropy based on a group that "you" assume is clean to do so.

I understand your sentiments, and they are absolutely valid.

Also, I would like you to tackle problems such as:

- multiple testing tooling
- migrations to different tools
- noise around the production source code
- where do you store helper/utilities related to those tools?
- add configurations that don't conflict with each other.

Reply:
Here's a  summary of a slack conversation: 

Keeping tests and stories grouped with source code is admittedly easier to navigate, but it comes at some risk.

1. Over time we build test utilities and because of proximity, someone could accidentally import one of those test utilities into production causing bugs.
2. Keeping tests and stories grouped by directory encourages a 1:1 relationship with source:tests and source:stories
3. Tackling
   - multiple testing tooling
   - migrations to different tools
   - noise around the production source code
   - where do you store helper/utilities related to those tools?
   - add tool configurations that don't conflict with each other.



**What's wrong with 1:1:1 source:test:stories?**

One could argue every export is a candidate for specs and every component is a candidate for a storybook entry (aka documentation).   This is not to say every export needs to be tested.  Exercising restraint and being a cost-effective is a skill and art form.  However, testing an export in a file with a matching filename creates an intuitive place for tests.

The counter argument here is that 1:1 encourages TDD & unit testing which results in forcing your team to maintain tests based on how you decided to solve a problem.  Ultimately the goal is to test a system; it is not to test how you resolve the system.  The stronger alternative to traditional unit tests are tests written from the user perspective testing the functionality of the system.


**What's wrong with an `e2e` folder for end-to-end tests?**

Your e2e suite could deploy a variety of competing testing tools to leverage the strength of each.  For example `cypress` is easy to use but doesn't support all browsers so you might write a few browser tests in `puppeteer`.

Maybe you started with `nightmare.js` and want to switch to `cypress` or go from `mocha` to `jest`.


The counter argument here is that if organizing tests by purpose and intent doesn't prevent you from describing multiple intentions.  For example to cover the listed cases, you could have `./e2e/`, `./e2e-legacy/` and `./e2e-ie-9/`